models/sipnet/R/read_restart.SIPNET.R

```diff
@@ -43,33 +55,84 @@ read_restart.SIPNET <- function(outdir, runid, stop.time, settings, var.names, p
+  if (length(idxs) > 0) {
+    last <- tail(idxs, 1)
+  } else {
+    prior <- which(as.Date(real_time) <= as.Date(stop.time))
+    if (length(prior) > 0) {
+      last <- max(prior)
+      PEcAn.logger::logger.warn("read_restart.SIPNET: no exact match for stop.time; using most recent prior step.")
+    } else {
+      last <- length(real_time)
+      PEcAn.logger::logger.warn("read_restart.SIPNET: no prior step exists; using final timestep in file.")
+    }
+  }
   
   #### PEcAn Standard Outputs
   if ("AbvGrndWood" %in% var.names) {
-    forecast[[length(forecast) + 1]] <- PEcAn.utils::ud_convert(ens$AbvGrndWood[last],  "kg/m^2", "Mg/ha")
-    names(forecast[[length(forecast)]]) <- c("AbvGrndWood")
+    # AbvGrndWood -> forecast (Mg/ha) and skip if invalid
+    val <- tryCatch(
+      PEcAn.utils::ud_convert(.safe_num(ens$AbvGrndWood[last]), "kg/m^2", "Mg/ha"),
+      error = function(e) NA_real_
+    )
+    forecast[[length(forecast) + 1]] <- if (length(val) == 1 && is.finite(val)) val else NA_real_
+    names(forecast[[length(forecast)]]) <- "AbvGrndWood"
+    if (!is.finite(forecast[[length(forecast)]][1])) {
+      PEcAn.logger::logger.warn("AbvGrndWood missing/invalid at restart; using NA.")
+    }
     
-    wood_total_C    <- ens$AbvGrndWood[last] + ens$fine_root_carbon_content[last] + ens$coarse_root_carbon_content[last]
-    if (wood_total_C<=0) wood_total_C <- 0.0001 # Making sure we are not making Nans in case there is no plant living there.
+    # robust wood fractions
+    abv <- .safe_num(ens$AbvGrndWood[last])
+    fr  <- .safe_num(ens$fine_root_carbon_content[last])
+    cr  <- .safe_num(ens$coarse_root_carbon_content[last])
     
-    params$restart["abvGrndWoodFrac"] <- ens$AbvGrndWood[last]  / wood_total_C
-    params$restart["coarseRootFrac"]  <- ens$coarse_root_carbon_content[last] / wood_total_C
-    params$restart["fineRootFrac"]    <- ens$fine_root_carbon_content[last]   / wood_total_C
-  }else{
-    params$restart["AbvGrndWood"] <- PEcAn.utils::ud_convert(ens$AbvGrndWood[last],  "kg/m^2", "g/m^2")
-    # calculate fractions, store in params, will use in write_restart
-    wood_total_C    <- ens$AbvGrndWood[last] + ens$fine_root_carbon_content[last] + ens$coarse_root_carbon_content[last]
-    if (wood_total_C<=0) wood_total_C <- 0.0001 # Making sure we are not making Nans in case there is no plant living there.
+    wood_total_C <- abv + fr + cr
+    if (!isTRUE(is.finite(wood_total_C) && wood_total_C > 0)) wood_total_C <- 1e-4
+    
+    params$restart["abvGrndWoodFrac"] <- abv / wood_total_C
+    params$restart["coarseRootFrac"]  <- cr  / wood_total_C
+    params$restart["fineRootFrac"]    <- fr  / wood_total_C
+    
+  } else {
+    # store AbvGrndWood (g/m^2) into params
+    val_gm2 <- tryCatch(
+      PEcAn.utils::ud_convert(.safe_num(ens$AbvGrndWood[last]), "kg/m^2", "g/m^2"),
+      error = function(e) NA_real_
+    )
+    if (length(val_gm2) == 1 && is.finite(val_gm2)) {
```

Comment:
Why did you check if the `length(val_gm2) == 1`? I don't think it's necessary because the `last` object has a length of one already. Not just this `if-control`, you probably want to check the issue throughout the script.